lightningnetwork/lnd
# WIP: Route Blinding [1/n]: TLV Record Definition and Blind Hop Processing

This (definitely draft) PR equips the `ChannelLink` to process HTLCs inside a blinded route. The bulk of the processing for blind hops in done by the *incoming* link (ie: the link on which the HTLC ADD arrives). The incoming link will: 
- Decrypt the onion using the ephemeral (route) blinding point if it is available on the *UpdateAddHTLC* TLV extension.
- Decrypt the route blinding payload and provide the forwarding information therein to the Switch.
- Compute the ephemeral (route) blinding point the next hop in the route will need to decrypt the onion and include that as a TLV extension to the *UpdateAddHTLC* message.

The *outgoing* link simply needs to forward the *UpdateAddHTLC* message it is given by the Switch. Everything will have been assembled for successful forwarding by the incoming link.

TODO: Several branches are still far too dirty to post publicly. Clean up branches for following:
- this branch
    - Consider more deeply the HTLC update retransmission (in-memory and from disk) code path & error handling.
- Route Blinding [2/n]: Dummy Hop Processing
    - May require some modifications to the HTLC Switch/Link (possibly channel state machine?) so that onions can be reprocessed.
- Route Blinding [3/n]: Blind Path Creation/Build Blind Route RPC
- Route Blinding [4/n]: Introduction Node and Blind Path Selection
    -  In attempting to define optimality and automate the discovery of a particular introduction node and blinded path, one reaches the point where the rubber seemingly really meets the road. 

### Comments
As of now this PR punts on dummy hop processing, integrating blinded routes with invoices, an algorithm for selecting the introduction node, and automated blinded route pathfinding. With this PR alone, LND won't be able to create or receive payment over a blinded route just yet, but it *will* be available for others to use while building blinded routes. Last I heard, a large portion of nodes on the network run LND, so even this modest start to an implementation could provide a boon to anonymity set sizes.

I tried to split things up such that the things I left out would not, if considered, fundamentally change the approach, though lacking a complete knowledge of BOLT-12 or Trampoline Routing, it is very possible that I did not succeed in this. 

The [route-blinding spec]() is still a living document, so I may be behind on some of the recent updates, particularly those on error handling. There is also a fair bit of cleanup left to do - far more than I would like for posting this at present - but after reading [this](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#share-early-share-often), I wanted to get this up sooner and get some information on where Lightning Labs is at in this area before spending too much more time deep in the weeds. 

## Steps to Test
- Took a stab at some automated tests at the `htlcswitch` package level.
    - go test -v -run ^TestChannelLinkBlindedPathPayment$ github.com/lightningnetwork/lnd/htlcswitch --tags=dev
- Have a local branch with an `itest` level which sets up a linear network `Alice <---> Bob <---> Carol <---> Dave`, creates a blinded route via RPC, concatenates the route to a normal `route.Route{}` and then verifies that payment can be sent successfully. Can post once it's a bit cleaner.
- TODO(11/20/22): Deploy a sim-net on Polar to test interoperability with other implementations: Use eclair or a further modified LND (which includes ability to build basic blinded routes) to generate a blinded route which routes across eclair/LND nodes, and see if the payment succeeds!

## Pull Request Checklist
### Testing

### Code Style and Documentation

📝 Please see our  [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.
